<a href="https://colab.research.google.com/github/marcelodepaoli/11_Regressao_Linear/blob/main/02_Projeto_de_Regress%C3%A3o_Linear.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Projeto de Regressão Linear

Parabéns! Você acabou de conseguir um contrato de trabalho com uma empresa de comércio eletrônico com sede em Nova York que vende roupas on-line, mas eles também têm sessões de aconselhamento sobre roupas e estilo na loja. Os clientes entram na loja, têm sessões/reuniões com um personal stylist, depois podem ir para casa e encomendar as roupas que desejam em um aplicativo móvel ou site.

A empresa está tentando decidir se deve concentrar seus esforços na experiência do aplicativo móvel ou no site. Eles contrataram você para ajudá-los a descobrir! Vamos começar!

Basta seguir as etapas abaixo para analisar os dados dos clientes (os dados são falsos, não são números de cartão de crédito ou e-mails reais).

##Imports
**Importe pandas, numpy, matplotlib e seaborn. Em seguida, defina %matplotlib inline (Você importará o sklearn conforme necessário.)**

##Obtenha os Dados

Trabalharemos com o arquivo csv de clientes de comércio eletrônico da empresa. o Arquivo tem informações do cliente como: e-mail, endereço e sua cor de avatar. Em seguida, também possui colunas de valor numérico:

* Avg. Session Length: Tempo médio das sessões de aconselhamento de estilo na loja
* Time on App: Tempo médio gasto no aplicativo em minutos
* Time on Website: Tempo médio gasto no site em minutos
* Length of Membership: Há quantos anos o cliente é membro 

**Leia o arquivo Ecommerce Customers csv como um DataFrame chamadp customers.**

In [ ]:
# Esse trecho do código é obrigatório para quem estiver fazendo tudo pelo colab
# Caso você esteja utilizando o jupyter pode comentar/apagar
import os
from google.colab import drive

drive.mount('/content/drive')
os.chdir("drive/My Drive/Colab Notebooks/11_Regressão_Linear")
os.listdir()

Mounted at /content/drive


['USA_Housing.csv',
 'Ecommerce Customers',
 '01 - Regressão Linear com Python.ipynb',
 '03 - Projeto de Regressão Linear - SOLUÇÃO.ipynb']

**Verifique o head() dos clientes e confira seus métodos info() e describe().**

## Exploratory Data Analysis

**Vamos explorar os dados!**

Para o resto do exercício, usaremos apenas os dados numéricos do arquivo csv.
___
**Utilize o seaborn para criar um joinplot para comparar as colunas Tempo no site e Valor anual gasto. A correlação faz sentido?**

In [ ]:
#sns.set_palette("GnBu_d")
#sns.set_style('whitegrid')
#sns.jointplot(x='Time on Website',y='Yearly Amount Spent',data=customers)

**Faça o mesmo, mas com a coluna Time on App.**

In [ ]:
#sns.jointplot(x='Time on App',y='Yearly Amount Spent',data=customers)

**Use jointplot para criar um gráfico hexadecimal 2D comparando o tempo no aplicativo e o comprimento da associação.**

In [ ]:
#sns.jointplot(x='Time on App',y='Length of Membership',kind='hex',data=customers)

**Vamos explorar esses tipos de relacionamentos em todo o conjunto de dados. Usando [pairplot](https://stanford.edu/~mwaskom/software/seaborn/tutorial/axis_grids.html#plotting-pairwise-relationships-with-pairgrid-and-pairplot) para recriar o enredo abaixo. (Não se preocupe com as cores)**

In [ ]:
#sns.pairplot(customers)

**Com base neste gráfico, qual parece ser o recurso mais correlacionado com o valor gasto anual?**

**Crie um gráfico de modelo linear (usando o lmplot da seaborn) do Yearly Amount Spent vs. Length of Membership.**

In [ ]:
#sns.lmplot(x='Length of Membership',y='Yearly Amount Spent',data=customers)

##Dados de Treinamento e Teste

Agora que exploramos um pouco os dados, vamos em frente e dividir os dados em conjuntos de treinamento e teste.

**Defina uma variável X igual às características numéricas dos clientes e uma variável y igual à coluna "Yearly Amount Spent".**

**Use model_selection.train_test_split do sklearn para dividir os dados em conjuntos de treinamento e teste. Defina test_size=0.3 e random_state=101**

##Treinando o Modelo

Agora é hora de treinar nosso modelo em nossos dados de treinamento!

**Importar LinearRegression de sklearn.linear_model**

**Crie uma instância do modelo LinearRegression() chamado lm.**

**Train/fit lm nos dados de treinamento.**

**Imprima os coeficientes do modelo**

##Previsão de Dados de Teste

Agora que ajustamos nosso modelo, vamos avaliar seu desempenho prevendo os valores de teste!

**Utilize lm.predict() para prever os dados do X_test set.**

**Crie um gráfico de dispersão dos valores de teste reais versus os valores previstos.**

In [ ]:
#plt.scatter(y_test,predictions)
#plt.xlabel('Y Test')
#plt.ylabel('Predicted Y')

##Avaliando o Modelo

Vamos avaliar o desempenho do nosso modelo calculando a soma dos quadrados dos resíduos e a pontuação da variância explicada (R^2).

**Calcule o erro médio absoluto, o erro quadrático médio e o erro quadrático médio. Consulte os slides ou a Wikipédia para as fórmulas**

##Resíduos

Você deve ter conseguido um modelo muito bom e com um bom ajuste. Vamos explorar rapidamente os resíduos para garantir que tudo esteja bem com nossos dados. 

**Plote um histograma dos resíduos e certifique-se de que parecem seguir uma distribuição normal. Use o displot seaborn (ou histplot) ou apenas plt.hist().**

In [ ]:
#sns.displot((y_test-predictions),bins=50,kde=True);

##Conclusão
Ainda queremos descobrir a resposta para a pergunta original, focamos nosso esforço no desenvolvimento de aplicativos para dispositivos móveis ou sites? Ou talvez isso nem importe, e o tempo de associação é o que realmente importa. Vamos ver se podemos interpretar os coeficientes para ter uma ideia.

**Recrie o dataframe abaixo.**

Coeffecient
Avg. Session Length     25.981550
Time on App             38.590159
Time on Website          0.190405
Length of Membership    61.279097

**Como você pode interpretar esses coeficientes?**

Interpretando os coeficientes:

- Mantendo todas as outras características fixas, um aumento de 1 unidade no **Avg. Session Length** está associado a um **aumento de 25.98 no total de dólares gastos**.
- Mantendo todas as outras características fixas, um aumento de 1 unidade no **Time on App** está associado a um **aumento de 38.59 no total de dólares gastos**.
- Mantendo todas as outras características fixas, um aumento de 1 unidade no **Time on Website** está associado a um **aumento de 0.19 no total de dólares gastos**.
- Mantendo todas as outras características fixas, um aumento de 1 unidade no **Length of Membership** está associado a um **aumento de 61.27 no total de dólares gastos**.

**Você acha que a empresa deveria se concentrar mais em seu aplicativo móvel ou em seu site?**

A resposta é complicada. Há duas maneiras de pensar sobre isso: desenvolver o site para acompanhar o desempenho do aplicativo móvel ou desenvolver mais o aplicativo, pois ele está funcionando melhor. 

Esse tipo de resposta realmente depende dos outros fatores que estão acontecendo na empresa. Seria mais prudente explorar a relação entre a duração da associação e o aplicativo ou o site antes de chegar a uma conclusão!